### ISSUE TITLE
Merge dupe layer styles with same name crashes Sketch

### ISSUE THREAD

**Issue by Bequiagal** [oodesign/merge-duplicate-symbols#113]
Issue: Merge duplicate layer styles with same name never finishes- it tries for 14 minutes or so and then crashes Sketch. 

• this happens whether I am only working on one document or if I am also including libraries and trying to merge between multiple libraries
• It also happens when I am only using the Merge Duplicates and Abstract plugins (with all others disabled). 
• If I am merging duplicate layer styles within one file, it will work if I merge just a few (less than 8) but it will take more than 15 minutes. 
• If I am merging even just one duplicate layer styles between libraries, the clock will spin away for 15 minutes followed by a crash. 

Here are my version details: 
MAC OS: 10.15.7
Sketch: 69.1
Merge Duplicates:  6.5.1

**Comment by oodesign** [comment]
Hi @Bequiagal ,

Sorry to hear it's not working fine for you... 😔. 

I'm afraid we cannot reproduce the issue you describe with our testing set, though. 

To investigate a bit further your issue, may you please send us the Merge Duplicates logs? Logs don't contain any sensitive information about your design file, it's just plugin processing related information. You can do so by:

1. Quit Sketch
2. Open Sketch again, and open the file you're having issues with
3. Merge Duplicates logs are disabled by default. You may enable them by going to Plugins > Merge duplicates... > Settings , and switching on the logging ability.
4. Run the Merge Duplicates command you were having issues with (Merge Symbols, Merge Text Styles or Merge Layer Styles, or any of their variants)
5. Proceed as usual until you experience the issue you described. **Don't quit Sketch yet** (this is important, as Sketch logs are deleted when you quit the app).
6. Logs will be stored by this point. May you please [send us](mailto:oscar@oodesign.me) your log file so that we can investigate what may be causing the issue? You may find them in ~/Library/Logs/com.bohemiancoding.sketch3/Plugin Log.log

Thanks a lot for your help! 🥰

**Comment by Bequiagal** [comment]
How does that work if Sketch quits?

--
Susy Mason
Senior UX  and Digital Designer

University of Vermont
University Communications
Johnson House
617 Main Street
Burlington, VT 05401

T: 802-656-7883

Visit our Drupal Web Guide for documentation and to learn about joining us in help hours for personal support.<http://www.uvm.edu/drupalwebguide>

**Comment by oodesign** [comment]
Hi! 

Actually logs are re-initialized when you launch Sketch again, so even if it quits (or is forced to quit) logs should remain there. 

You can also open the PluginLog.log file in any text editor and then run Merge Duplicates. If you do you should see the logging process live. When the plugin crashes (or Sketch quits) it will stop. If you send us the file at that point (or copy us its content) it'd be fine.

**Comment by Bequiagal** [comment]
Here you go.

Mac Crash Report and the log file after Sketch crashed when trying to Merge duplicates with same name across libraries. I was just trying to do 4 merges…. It went for 13 minutes and then crashed.

Thank you for your help

Susy


--
Susy Mason
Senior UX  and Digital Designer

University of Vermont
University Communications
Johnson House
617 Main Street
Burlington, VT 05401

T: 802-656-7883

Visit our Drupal Web Guide for documentation and to learn about joining us in help hours for personal support.<http://www.uvm.edu/drupalwebguide>

**Comment by oodesign** [comment]
Hi @Bequiagal ,

I'm afraid GitHub doesn't add the attachments when the reply is sent via mail.
May you attach it manually, or [send us](mailto:oscar@oodesign.me) the logs via mail?

Thank you!

**Comment by Geomaxi** [comment]
Hi there, 
I thought I'd join in, cause I am experiencing the same issue with "Merge layer styles (from list)" and "Merge text styles (from list).

I want to merge local layer and text styles with ones from the library. My sketch file is quite big (330+ artboards, so I assume a couple of thousand instances of some layer styles.). I know that's a lot, but the files have grown over time and we didn't have your amazing plugin before
I was able to merge layer styles that weren't used so frequently, but the ones that are used a lot, crash the App and my Macbook sound like it is about to take off.
Once my Mac even showed me a warning (while Merge duplicates was running), that I was running out of storage, cause Sketch was using 130 GB of storage at that time. 

Unfortunately I can't share my sketch files with you (project for a client).

Besides that I also have another issue and a wish/feature reqeust: 

1. When opening the Merge layer styles window, it shows the same layer style from the library multiple times on the side (more than 20 times). But when I open the library the style exists only once. ![Same layer style shown multiple times](https://user-images.githubusercontent.com/73465467/97151466-d490cc00-176f-11eb-93d0-79a1babadbd6.png)

2. A feature request would be if it is possible to show some sort of progress bar/loading indicator, when the plugin merges styles, cause it would be amazing to see how long it would approx. take, instead of the Mac's spinning wheel.

Thank you, 
Max

MacOS Catalina 10.15.7;
Sketch 69.1; 
Merge duplicates 6.5.2;

**Comment by Bequiagal** [comment]
My experience is similar-- I love this plugin but I think in my case, Merge Duplicates is getting confused somehow by the Abstract branches.
 
Take a look at this scenario: 

• One Branch open in Abstract
• Run "Merge Duplicate text styles with the same name"

Results: 
It finds (among other things) 3 instances of 
dt/p/p/darkGray/left
 
… and they are all in the SAME DOCUMENT (libraries not included)

BUT if you navigate to the new Components/Text Styles/ you can see there is only ONE layer style with this name.
 
If I try to merge to one, it crashes… I think because there actually _are not more than one_.
 
I have included a marked up screen grab here- log files sent directly.
 
Thank you for looking into this- so appreciate it.


![3Txtstyles-sameName-Found](https://user-images.githubusercontent.com/49312665/97173632-4284c580-1767-11eb-96d0-9fa225cc4319.png)

 
Susy

**Comment by oodesign** [comment]
Hi @Geomaxi  and @Bequiagal ,

Thank you for taking the time to send us this thorough feedback. It helps a lot!

Looks like Merge Duplicates is considering as duplicates some styles that shouldn't. As you point out, it may be related with how Abstract handles linked libraries, and Sketch foreign styles related with them.

We'll investigate this issue and come back to you when we release a fix related with it.

Thank you!

**Comment by Bequiagal** [comment]
Thank you-- let me know if you need anything from me. I'll submit the same issue to Abstract and report back findings.

**Comment by Geomaxi** [comment]
Thank you very much. I sent you my crash reports via mail as well.

Btw I am not using Abstract or any other versioning tool (I manually create versions of the file and store them in a separate folder, by duplicating/renaming it every now and then). 
We only use Zeplin to share the designs, but I can't imagine it has anything to do with that.

**Comment by Bequiagal** [comment]
This may provide a wee bit more information. 

In this instance, I am showing "Merge Duplicate Text Styles" but only on the current document with "Libraries" turned off. 

The current document is the ds__typography library document. 

Merge Duplicate Text styles thinks the type style is duplicated as it is in the current file and the branch of the current file. 
In both cases, we are still just talking about the same text style and the same document as (This Sketch File) and the Typographylibrary file are the same. 

![typography-is-current](https://user-images.githubusercontent.com/49312665/97182291-bed0d600-1772-11eb-9493-1c893c6bc870.png)

**Comment by oodesign** [comment]
Thanks for these further details!

@Bequiagal , in this last capture Merge Duplicates is finding two different styles with the same name. One coming from the "ds_typography (Merge Duplicates- document bug for dupe layer styles)", and another one in the local file. However, per what you describe, does this style exist only in one of the files? If so, may I assume that the style is coming from the linked library?

**Comment by Bequiagal** [comment]
Both of those documents are one in the same-- and I believe there is only 1 text style--- but Merge Duplicates thinks there are 2- one for the document and one for the branch of the same document. (because all Sketch docs are opened through Abstract). 

Here's another view into this problem: 
(attached) 
If i go into Sketch/Preferences/Libraries
you can see that Sketch thinks that each branch -- including old ones that have been merged and no longer are branched - are still active libraries. 

I think this is an important clue to the issue I am experiencing. The document name is followed by the (branch name) in parenthesis.

![ancient-libraries](https://user-images.githubusercontent.com/49312665/97306882-69113200-1835-11eb-8f10-49a297e0cc9b.png)

**Comment by Bequiagal** [comment]
I think the root of this issue had to do with Sketch holding on somehow to old library branches that had since been merged but were still showing up as linked libraries (duplicates actually, as they were closed branches of currrently branched libraries). 

I reduced plugins to Merge duplicates, Rename it, Abstract and the 2 InVision plugins, bravely just deleted these old branches from Sketch/Prefs/Libraries- and everything seems fine and tidy so far and working as expected. 

So again, I don't think it was a Merge Duplicates issue at all-- but I am not sure what caused Sketch to hold on to these old library branches. Merge Duplicates was looking at phantom library data. 

I'm OK with closing this issue out.

**Comment by oodesign** [comment]
HI @Bequiagal ,

Thank you for looking deep into this and sharing this info with us. I'm glad you did find a workaround for it.

@Geomaxi, you seem to be having this very same issue (false duplicates) without using Abstract. May I ask you to check if a similar issue may be happening in your case (libraries that shouldn't be there appear as active in the Sketch Preferences > Libraries window)?

Thanks!

**Comment by Geomaxi** [comment]
Hi, I just checked and there is only the one library in the window. In Finder there is also only this version of the library that I want to merge with. So it might be something else in my case. 
Also I have been able to merge styles that were not as frequently used in the sketch file, but the ones that are used a lot crash the app when I'm trying to merge them with the library (Same with text styles).

![Bildschirmfoto 2020-10-28 um 14 07 20](https://user-images.githubusercontent.com/73465467/97440187-6db61300-1927-11eb-883f-884bf2dfb82a.png)

**Comment by Geomaxi** [comment]
@oodesign Hi, I wanted to share an interesting observation I have made today. Yesterday I have moved a couple of new symbols to my library. Of course many of them use the layer styles that are defined in the library as well. 

I realized that when I open the Merge layer styles window now in my working sketch file (that uses the library), there are even more duplicated Layer styles, even though the layer style exists only once in the library. I guess the plugin counts every instance where a layer style is applied in the library and lists it as if it was a separate layer style.

For example: I have the color Primary 100. This color is defined as a layer style in the library once (see two screenshots of the layer styles captured in the library and in my working sketch file).
So it should only be displayed once in the merge layer styles window, right? But instead, what happens is, that it is displayed well over 40 times in the list (see screenshot of Merge duplicates window).
I seems to me that the reason is, that I also have symbols in my library that use the layer style Primary 100 and the plugin acts like every shape where Primary 100 is applied is a separate layer style and lists it, even though it is only defined once (same for all my other layer styles).

![SketchFile_LayerStyles](https://user-images.githubusercontent.com/73465467/98363219-e8211a00-202e-11eb-9888-26844e3bbf55.png)
![Library_LayerStyles](https://user-images.githubusercontent.com/73465467/98363225-eb1c0a80-202e-11eb-9037-bd228835489d.png)
![Merge Layer Styles from List Duplicates](https://user-images.githubusercontent.com/73465467/98363226-ebb4a100-202e-11eb-9152-7336a127498f.png)

**Comment by oodesign** [comment]
Hi @Geomaxi ,

Thanks for sharing these details with us!

Merge Duplicates operates with layer styles collections that Sketch stores in every document (not with layers or symbols).
The issue you describe is caused because, for some reason, the document is storing different "versions" (or references) of the same style. 

We are working to figure out in which situations this may happen, and which of those styles references would be "the right one" to use and merge towards in the plugin.

I'll make sure to keep you updated on it 😊

Thank you!

**Comment by hiyaneko** [comment]
I'm experiencing the same issue with merging styles (for layers and text) Sketch crashes everytime I try to merge styles and it takes too long to respond. I'm using a big file attached to an abstract library. It is also showing me multiple repeated styles coming from the same place.

**Comment by hiyaneko** [comment]
<img width="1617" alt="Screenshot 2020-11-17 at 10 07 01" src="https://user-images.githubusercontent.com/32107480/99369436-b759a400-28bc-11eb-8fc4-4011c8d2ae63.png">
this might help? I'm still having the same issue, sketch is crashing and the merge text/layer styles is not working still

**Comment by johnpettengill** [comment]
I am also experiencing these issues

- When I attempt to merge styles/symbols with the same name.... I see library provided styles/symbols repeated 5 or 6 times. (I use abstract, when I look for linked libraries I only see my one library from abstract branch)
- Sketch crashes

this makes this plugin absolutely useless. is this going to be fixed?

**Comment by oodesign** [comment]
Hi @johnpettengill ,

Absolutely. We're actively working with Sketch on a fix for it. 

I'll make sure to keep you updated on it.

**Comment by oodesign** [comment]
Hi!

We've identified a performance issue that will cause Merge Duplicates (and Sketch) to crash when there are a lot of symbol instances in the design file (as processing all of their overrides when merging symbols/styles is taking a huge amount of resources, causing Sketch to crash).

We continue to work with Sketch on a fix. I'll keep you updated on it.

**Comment by Bequiagal** [comment]
Thanks!

**Comment by oodesign** [comment]
Hi @johnpettengill @hiyaneko @Bequiagal @Geomaxi ,

We just released v8.1.0, that should be way faster processing **symbols** than the previous versions.
Bad news is that Layer & text styles may still be slow (for now). Good news is that... we're already applying the same improvements to style processing 👏,👏,👏,.

So... watch this space! 😊

**Comment by Bequiagal** [comment]
That's great! I'll check it out.

**Comment by johnpettengill** [comment]
Checked it out and symbols work like a dream - quick and snappy. Still a couple weird symbol copies coming up when using external libraries but I can live with that - don’t seem to have any big consequences.

Kudos on figuring it out, and looking forward to faster layer style merging!

**Comment by oodesign** [comment]
Hi @johnpettengill @hiyaneko @Bequiagal @Geomaxi ,

We just released v8.2.0, which should be way way faster processing layer and text styles.
May you please give it a try and let us know hoe it worked for you?

Thanks! 🥰

**Comment by johnpettengill** [comment]
Love seeing the progress indicator, it took a bit of time but it worked! Meanwhile it updated over 20,000 instances of colors - holy shit.

Y’all really did it!

Thanks so much, you guys are life savers and I so appreciate these fixes. Kudos!

**Comment by johnpettengill** [comment]
Yikes - this does not work.

All my style overrides for those key colors are now royally messed.



Some of them look correct, but then when I click on them I see that they actually aren’t mapping to the library symbol. They’re mapping to this “default style”. There are cases where I want it to be this color, so that’s non-ideal but it doesn’t APPEAR broken.

Then there are instances where it’s applying this “Default Style” and that default style is the wrong color.

Wow, this is… hours of work ahead of me to fix this.

**Comment by oodesign** [comment]
Hi @johnpettengill ,

Oh, so sorry to hear that 😔 .

This should definitely not happen. For some reason Merge Duplicates is able to apply the style itself, but not able to establish the reference (the "link") to it.

However, I'm afraid we're not able to reproduce the issue with our testing set. 

In order to investigate further the issue, may I ask you to please [send us](mailto:oscar@oodesign.me) your design file (and related libraries)?

Your files are and will remain confidential. We ensure they will only be used to investigate the issue you're describing, deleted after closing it, and never stored or used for any other means.

Also please feel free to use the sharing platform you prefer to send us the files, if you don't want to send them via email. Just [let us know](mailto:oscar@oodesign.me) the details to get them.

Thanks a lot!

**Comment by oodesign** [comment]
Hi @johnpettengill ,

Are you still experiencing issues with Merge Duplicates? Is there anything we can help with?

**Comment by Geomaxi** [comment]
Hi @oodesign, 
I've used the latest version of Merge duplicates and first of all I'd like to thank you a lot for the support and work you do on the plugin. I was finally able to merge all my old layer and text styles and I really love the loading indicator!  🥳

My experience was still a bit mixed though. For disclosuer I was using Merge duplicates 8.2.2 and Sketch 69.2 (I haven't update to Sketch 70 on my work laptop, cause I have heard that others were experiencing some issues with it and I am also not on Big Sur so I didn't take the risk yet.).

- The features **"Merge layer styles (from list)"** and **"Merge text styles (from list)"** worked perfectly. All the styles were merged correctly as far as  I can see and the performance was lightyears better than before.
- The only issue I had was that there were still some text styles from my library that were displayed multiple times. That was a little confusing. I just ignored these though selected one of them and it merged just fine. Also those multiple styles weren't displayed in the actual library nor in the text/layer styles panels. (see screenshots) They're basically ghosts :-) 

![Sketch text styles panel](https://user-images.githubusercontent.com/73465467/105515167-a706e080-5cd4-11eb-80f5-420d200fd1c5.png)

![Merge duplicates text styles (from list) window](https://user-images.githubusercontent.com/73465467/105515177-aa01d100-5cd4-11eb-97cc-505a95eab239.png)


- I  also tried some other ways to merge the styles first (**"Merge layer styles with same name"** and **"Merge similar text styles"**), that seemed to be a little more efficiently than merging every style separately from the list. Those other features  did not work though. They crashed the app like before and the performance was very slow.

**Comment by oodesign** [comment]
Hi @Geomaxi ,

Thanks for taking the time to share these details with us. They help a lot!

### About duplicate ("ghost") styles

Every time a style is used in Sketch, a new reference to that style is saved in the document (internally). What means that on a given point, the document may have saved multiple references to a very same style. This typically doesn't last long, as Sketch itself "purges" the document, and keeps only the needed ones.

However, there are some situations (sudden crashes, early saves, ...) that could cause the document to get stored before having the time to purge it. And in those cases, for that document, several references to the same style may still be present.

We do our best to identify these "duplicated" references and to unify most of them, so that they're only presented once. Also, when merging, the "duplicated" references are also updated, so somehow we're also cleaning and purging the document. Though, there are some situations where we just can't identify if those references do belong to the same style, and therefore may still be presented many times in the merge UI. 

That said, we'll continue to work on it to, hopefully, not display duplicate references in the near future.

### About both crashes you describe

Obviously this is not expected 😅. May you please send us the plugin logs to investigate it further? Logs don't contain any sensitive information about your design file, it's just plugin processing related information. You can do so by:

1. Quit Sketch
2. Open Sketch again, and open the file you're having issues with
3. Merge Duplicates logs are disabled by default. You may enable them by going to Plugins > Merge duplicates... > Settings , and switching on the logging ability.
4. Run the Merge Duplicates command you were having issues with (Merge Symbols, Merge Text Styles or Merge Layer Styles, or any of their variants)
5. Proceed as usual until you experience the issue you described. **Don't quit Sketch yet** (this is important, as Sketch logs are deleted when you quit and relaunch the app).
6. Logs will be stored by this point. May you please [send us](mailto:support@oodesign.me) your log file so that we can investigate what may be causing the issue? You may find them in ~/Library/Logs/com.bohemiancoding.sketch3/Plugin Log.log

Thanks a lot for your help! 🥰
